lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskCommentServiceImpl.java

```diff
@@ -0,0 +1,279 @@
+import pro.taskana.task.api.models.TaskComment;
+import pro.taskana.task.internal.models.TaskCommentImpl;
+
+class TaskCommentServiceImpl {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TaskCommentServiceImpl.class);
+
+  private static final String NOT_AUTHORIZED =
+      " Not authorized, TaskComment creator and current user must match. TaskComment creator is ";
+  private static final String BUT_CURRENT_USER_IS = " but current user is ";
+  private static final String ID_PREFIX_TASK_COMMENT = "TCI";
+  private InternalTaskanaEngine taskanaEngine;
+  private TaskServiceImpl taskService;
+  private TaskCommentMapper taskCommentMapper;
+
+  TaskCommentServiceImpl(
+      InternalTaskanaEngine taskanaEngine,
+      TaskCommentMapper taskCommentMapper,
+      TaskServiceImpl taskService) {
+    super();
+    this.taskanaEngine = taskanaEngine;
+    this.taskService = taskService;
+    this.taskCommentMapper = taskCommentMapper;
+  }
+
+  TaskComment updateTaskComment(TaskComment taskCommentToUpdate)
+      throws NotAuthorizedException, ConcurrencyException, TaskCommentNotFoundException,
+          TaskNotFoundException {
+
+    LOGGER.debug("entry to updateTaskComment (taskComment = {})", taskCommentToUpdate);
+
+    String userId = CurrentUserContext.getUserid();
+
+    TaskCommentImpl taskCommentImplToUpdate = (TaskCommentImpl) taskCommentToUpdate;
+
+    try {
+
+      taskanaEngine.openConnection();
+
+      taskService.getTask(taskCommentImplToUpdate.getTaskId());
+
+      if (taskCommentToUpdate.getCreator().equals(userId)) {
+
+        TaskCommentImpl oldTaskComment = getTaskComment(taskCommentImplToUpdate.getId());
+
+        checkModifiedHasNotChanged(oldTaskComment, taskCommentImplToUpdate);
+
+        taskCommentImplToUpdate.setModified(Instant.now());
+
+        taskCommentMapper.update(taskCommentImplToUpdate);
+
+        LOGGER.debug(
+            "Method updateTaskComment() updated taskComment '{}' for user '{}'.",
+            taskCommentImplToUpdate.getId(),
+            userId);
+
+      } else {
+        throw new NotAuthorizedException(
+            NOT_AUTHORIZED + taskCommentImplToUpdate.getCreator() + BUT_CURRENT_USER_IS + userId,
+            userId);
+      }
+    } finally {
+      taskanaEngine.returnConnection();
+      LOGGER.debug("exit from updateTaskComment()");
+    }
+
+    return taskCommentToUpdate;
+  }
+
+  TaskComment createTaskComment(TaskComment taskCommentToCreate)
+      throws NotAuthorizedException, TaskCommentAlreadyExistException, TaskNotFoundException {
+
+    LOGGER.debug("entry to setTaskComment (taskCommentToCreate = {})", taskCommentToCreate);
+
+    TaskCommentImpl taskCommentImplToCreate = (TaskCommentImpl) taskCommentToCreate;
+
+    try {
+
+      taskanaEngine.openConnection();
+
+      taskService.getTask(taskCommentImplToCreate.getTaskId());
+
+      if (isTaskCommentAlreadyExisting(taskCommentImplToCreate.getId())) {
```

Comment:
if the Id is already set on createTaskComment call, an  InvalidArgumentException should be thrown.
Please modify createTask correspondingly...

Reply:
Done